service/api/service.go

```diff
@@ -156,15 +156,30 @@ func (s *serviceImpl) ApiV1WorkflowWaitForStateCompletion(
 ) (wresp *iwfidl.WorkflowWaitForStateCompletionResponse, retError *errors.ErrorAndStatus) {
 	defer func() { log.CapturePanic(recover(), s.logger, &retError) }()
 
-	var workflowId string
+	var parentWfId, currentWorkflowId, _ string
+
+	response, err := s.client.DescribeWorkflowExecution(ctx, req.GetWorkflowId(), "", nil)
+	if err != nil {
+		return nil, s.handleError(err, WorkflowWaitForStateCompletionApiPath, req.WorkflowId)
+	}
+
+	if response.FirstRunId == "" {
+		parentWfId = req.WorkflowId // Cadence
+	} else {
+		parentWfId = response.FirstRunId // Temporal
+	}
+
 	if req.WaitForKey != nil {
-		workflowId = service.IwfSystemConstPrefix + req.WorkflowId + "_" + *req.StateId + "_" + *req.WaitForKey
+		currentWorkflowId = service.IwfSystemConstPrefix + req.WorkflowId + "_" + *req.StateId + "_" + *req.WaitForKey
+		_ = service.IwfSystemConstPrefix + parentWfId + "_" + *req.StateId + "_" + *req.WaitForKey
 	} else {
-		workflowId = service.IwfSystemConstPrefix + req.WorkflowId + "_" + *req.StateExecutionId
+		currentWorkflowId = service.IwfSystemConstPrefix + parentWfId + "_" + *req.StateExecutionId
+		_ = service.IwfSystemConstPrefix + parentWfId + "_" + *req.StateExecutionId
 	}
 
 	options := uclient.StartWorkflowOptions{
-		ID:        workflowId,
+		// TODO: Switch currentWorkflowId to _ (after renaming) after a short amount of time to ensure backward compatibility
```

Comment:
Ohh interesting! I was thinking of making two deployment -- first deploy the workflowImpl code (which signalWIthStart both), after some time, deploy the api service. 
But what you put here made me feel like it's nice to use a config? 
Potentially using the `Api ApiConfig `yaml:"api"`` in https://github.com/indeedeng/iwf/blob/main/config/config.go 
```
waitForStateCompletionMigration:
   SignalWithStartOn: both/new/old
   WaitForOn:old/new
```

Reply:
In workflowImpl, we have access to ApiConfig via env like this: https://github.com/indeedeng/iwf/blob/main/service/interpreter/continueAsNewer.go#L57

Reply:
That's a good idea. I'll implement it

Reply:
Done